File: src/main/java/com/leijendary/spring/microservicetemplate/error/ResourceNotUniqueExceptionHandler.java
package com.leijendary.spring.microservicetemplate.error;

import com.leijendary.spring.microservicetemplate.data.response.ErrorResponse;
import com.leijendary.spring.microservicetemplate.exception.ResourceNotUniqueException;
import lombok.RequiredArgsConstructor;
import org.springframework.context.MessageSource;
import org.springframework.core.annotation.Order;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import static com.leijendary.spring.microservicetemplate.util.RequestContextUtil.getLocale;
import static org.springframework.http.HttpStatus.CONFLICT;

@RestControllerAdvice
@Order(3)
@RequiredArgsConstructor
public class ResourceNotUniqueExceptionHandler {

    private final MessageSource messageSource;

    @ExceptionHandler(ResourceNotUniqueException.class)
    @ResponseStatus(CONFLICT)
    public ErrorResponse catchResourceNotUnique(final ResourceNotUniqueException exception) {
        final var code = "validation.alreadyExists";
        final var message = messageSource.getMessage(code,
                new Object[] { exception.getField(), exception.getValue() }, getLocale());

        return ErrorResponse.builder()
                .addError(exception.getField(), code, message)
                .status(CONFLICT)
                .build();
    }
}

File: src/main/kotlin/com/leijendary/error/ResourceNotUniqueExceptionHandler.kt
package com.leijendary.error

import com.leijendary.error.exception.ResourceNotUniqueException
import com.leijendary.model.ErrorModel
import com.leijendary.util.requestContext
import org.springframework.context.MessageSource
import org.springframework.core.annotation.Order
import org.springframework.http.HttpStatus.CONFLICT
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
@Order(3)
class ResourceNotUniqueExceptionHandler(private val messageSource: MessageSource) {
    @ExceptionHandler(ResourceNotUniqueException::class)
    @ResponseStatus(CONFLICT)
    fun catchResourceNotUnique(exception: ResourceNotUniqueException): List<ErrorModel> {
        val code = "validation.alreadyExists"
        val source = exception.source
        val field = source.pointer?.split("/")?.last()
        val arguments = arrayOf(field, exception.value)
        val message = messageSource.getMessage(code, arguments, requestContext.locale)
        val error = ErrorModel(id = exception.value.toString(), code = code, message = message, source = source)

        return listOf(error)
    }
}
